projects/plugins/protect/src/js/components/scan-onboarding/index.jsx

```diff
@@ -0,0 +1,97 @@
+import { ActionPopover } from '@automattic/jetpack-components';
+import { useMemo, useEffect } from 'react';
+import useOnboarding from '../../hooks/use-onboarding';
+import useProtectData from '../../hooks/use-protect-data';
+import useThreatsList from '../threats-list/use-threats-list';
+import useScanPopoverArgs from './use-scan-popover-args';
+
+const ScanOnboarding = ( { anchors } ) => {
+	const { freeScanOnboardingDismissed, paidScanOnboardingDismissed } =
+		window.jetpackProtectInitialState;
+	const { hasRequiredPlan } = useProtectData();
+	const { onboardingStep, getCurrentPopoverArgs, resetOnboardingOnAnchorRegeneration } =
+		useOnboarding();
+
+	const { list } = useThreatsList();
+	const fixableList = list.filter( obj => obj.fixable );
+
+	const maxSteps = 4;
+	const minSteps = 2;
+	const alternateMaxSteps = 3;
+
+	// Define logic for determining total steps
+	const calculateTotalSteps = () => {
+		if ( ! hasRequiredPlan || list.length === 0 ) {
+			return minSteps;
+		}
+		if ( fixableList.length === 0 ) {
+			return alternateMaxSteps;
+		}
+		return maxSteps;
+	};
+
+	const totalSteps = useMemo( calculateTotalSteps, [ hasRequiredPlan, list, fixableList ] );
+
+	// Retrieve args defined for each popover
+	const popoverArgs = useScanPopoverArgs( { anchors, totalSteps } );
+
+	// Define logic for determining which popover args to use
+	const scanOnboardingStepHandlers = {
+		// Handle step 1
+		1: () => {
+			if (
+				( ! hasRequiredPlan && ! freeScanOnboardingDismissed ) ||
+				( hasRequiredPlan && ! paidScanOnboardingDismissed )
+			) {
+				return popoverArgs.yourScanResults;
+			}
+		},
+		// Handle step 2
+		2: () => {
+			if ( ! hasRequiredPlan && ! freeScanOnboardingDismissed ) {
+				return popoverArgs.dailyAutomatedScans;
+			} else if ( ! paidScanOnboardingDismissed ) {
+				if ( list.length === 0 ) {
+					return popoverArgs.dailyAndManualScans;
+				} else if ( fixableList.length === 0 ) {
+					return popoverArgs.understandSeverity;
+				}
+				return popoverArgs.fixAllThreats;
+			}
+		},
+		// Handle step 3
+		3: () => {
+			if ( hasRequiredPlan && ! paidScanOnboardingDismissed ) {
+				return fixableList.length === 0
+					? popoverArgs.dailyAndManualScans
+					: popoverArgs.understandSeverity;
+			}
+		},
+		// Handle step 4
+		4: () => {
+			if ( hasRequiredPlan && ! paidScanOnboardingDismissed ) {
+				return popoverArgs.dailyAndManualScans;
+			}
+		},
+	};
+
+	const scanOnboardingPopoverArgs = useMemo( () => {
+		return getCurrentPopoverArgs( scanOnboardingStepHandlers );
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
In general exhaustive deps is important to ensure you're doing things The React Way, so the disables make me suspicious that the problem this memoization is solving might be avoidable higher up somehow. Could this value be provided directly from the onboarding hook instead?

Reply:
>In general exhaustive deps is important to ensure you're doing things The React Way, so the disables make me suspicious that the problem this memoization is solving might be avoidable higher up somehow.

I see we've previously disabled the eslint rule to avoid unnecessary re-renders and/or if we've wanted to remove items from the dep array that we're certain will not change. In this particular case, based on the actions of `getCurrentPopoverArgs`and `scanOnboardingStepHandlers`, I don't believe adding or keeping them removed will make any significant difference, but I agree that we should follow The React Way and only disable the rule where necessary.

>Could this value be provided directly from the onboarding hook instead?

I structured it this way initially to prioritize reusability. The intention was for `getCurrentPopoverArgs` to be a shared method that takes in a list of directives for how and when to return certain args, and determines which one to display based on the current state of `onboardingStep`, among other variables (`hasRequiredPlan`, `selected`, etc). My thought was that eventually, we may have another set of onboarding steps/rules for the WAF perhaps, we can develop the logic specific to those steps and pass them to this function to determine which one should be returned when, then pass that value to an `ActionPopover` component for display. That is why I have separated the logic between the hook where we have the shared methods and a component where we have methods and data specific to those onboarding flows, however, I am open to adjusting that approach to centralize things if we have no intention of reusing this.